### ISSUE TITLE
Caching leads to InvalidManifestError on self-hosted Azure DevOps agents

### ISSUE THREAD

**Issue by martin-schulze-e2m** [pre-commit/pre-commit#3347]
As also described by @sdebruyn in https://github.com/pre-commit/pre-commit/issues/2012#issuecomment-1195258944_
            
the [caching example](https://github.com/asottile-archive/azure-pipeline-templates/blob/main/job--pre-commit.yml) does not work on Azure DevOps with self hosted agents. The cache directory path seems to change per agent but the cache is global, leading to mismatches between the cache contents and its actual location as visible in the following error message:

```
pre-commit run --all-files --show-diff-on-failure
========================== Starting Command Output ===========================
/usr/bin/bash --noprofile --norc /__w/_temp/b381500a-2d6a-4979-8ad5-e90aad931288.sh
An error has occurred: InvalidManifestError: 
=====> /__w/60/pre-commit-cache/repouhpd7bak/.pre-commit-hooks.yaml is not a file
Check the log at /__w/1/pre-commit-cache/pre-commit.log
```
Note the `/__w/60` (probably comes from in the cached db) in the error message vs the `/__w/1` (the actual location) in the log. (This is a container based job, so the paths don't start with something like `/home/azure-agent/...`)

The script is slightly adapted but the problem also exists with the original:

```
jobs:
- job: #...
  steps:
    - script: echo "##vso[task.setvariable variable=PY]$(python -VV)"
      displayName: set version variables

    - task: Cache@2
      inputs:
        key: pre-commit | "4" | .pre-commit-config.yaml | "$(PY)"
        path: $(Pipeline.Workspace)/pre-commit-cache

    - script: pre-commit run --all-files --show-diff-on-failure
      displayName: run pre-commit
      env:
        PRE_COMMIT_HOME: $(Pipeline.Workspace)/pre-commit-cache
```

As I understand the comments in #2012, we have to provide a constant `PRE_COMMIT_HOME` to not run into this problem.

**Comment by asottile** [comment]
unless you have a suggestion there's nothing I can do about this.  the directions are aimed at the hosted service and it sounds like your self host is incompatible

**Comment by martin-schulze-e2m** [comment]
I suggest to add a caveat notice on your Azure DevOps example, that the caching does not work on self-hosted agents (due to the changing directories, similarly to GitLab). 

I also wanted to document the problem here in a searchable manner to help future readers. (And maybe gather solutions or at least discussions towards a solution.)

Obviously, I also hoped that you would have a workaround but since this is rooted in an Azure DevOps issue, I will have to direct further questions to them.

**Comment by martin-schulze-e2m** [comment]
I worked around the issue by not using the `Cache`  task at all. This means, the cache won't be shared between different agents. Instead, the local workspace of each agent is used. This is persistent for our self hosted agents.

I also tried to use the `Cache` task but provide `PRE_COMMIT_HOME=/pre-commit-cache` wich is symlinked to `$(Pipeline.Workspace)/pre-commit-cache`. However, pre-commit still stored the actual path instead of the symlinked one, so that did not work either.
